snow/engine/common/timer_test.go

```diff
@@ -0,0 +1,116 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package common
+
+import (
+	"sync"
+	"testing"
+	"time"
+)
+
+func TestTimeoutScheduler(t *testing.T) {
+	for _, testCase := range []struct {
+		expectedInvocationCount int
+		desc                    string
+		shouldPreempt           bool
+		clock                   chan time.Time
+		initClock               func(chan time.Time)
+		advanceTime             func(chan time.Time)
+	}{
+		{
+			desc:                    "multiple pendingTimeoutToken one after the other with preemption",
+			expectedInvocationCount: 10,
+			shouldPreempt:           true,
+			clock:                   make(chan time.Time, 1),
+			initClock:               func(chan time.Time) {},
+			advanceTime:             func(chan time.Time) {},
+		},
+		{
+			desc:                    "multiple pendingTimeoutToken one after the other",
+			expectedInvocationCount: 10,
+			clock:                   make(chan time.Time, 1),
+			initClock: func(clock chan time.Time) {
+				clock <- time.Now()
+			},
+			advanceTime: func(clock chan time.Time) {
+				clock <- time.Now()
+			},
+		},
+	} {
+		t.Run(testCase.desc, func(_ *testing.T) {
+			// Not enough invocations means the test would stall.
+			// Too many invocations means a negative counter panic.
+			var wg sync.WaitGroup
+			wg.Add(testCase.expectedInvocationCount)
+
+			testCase.initClock(testCase.clock)
+
+			var preemptionSignal PreemptionSignal
+			ps := preemptionSignal.Listen()
+
+			if testCase.shouldPreempt {
+				preemptionSignal.Preempt()
+			}
+
+			// Order enforces timeouts to be registered once after another,
+			// in order to make the tests deterministic.
+			order := make(chan struct{})
+
+			newTimer := makeMockedTimer(testCase.clock)
+
+			onTimeout := func() {
+				order <- struct{}{}
+				wg.Done()
+				testCase.advanceTime(testCase.clock)
+			}
+
+			ts := NewTimeoutScheduler(onTimeout, ps)
+			ts.newTimer = newTimer
+
+			for i := 0; i < testCase.expectedInvocationCount; i++ {
```

Comment:
Not asking to change this, but just pointing out that now that we are using go `1.22` this can be written as:
```suggestion
			for range testCase.expectedInvocationCount {
```

Reply:
I don't want to be the first to break incompatibility with 1.21 :-)